### ISSUE TITLE
Temporal Fusion Transformer implementation

### ISSUE THREAD

**Issue by Attol8** [timeseriesAI/tsai#206]
Hi, 

I would like to implement Time Fusion Transformers ([link](https://paperswithcode.com/paper/temporal-fusion-transformers-for#code)) used for forecasting in tsai. I am already familiar with fastai and would like to do so as a coding exercise to get more experience with Transformer based architectures. Also, the model is not implemented in tsai and would be a great addition due to its performance. Lots of code that I can use for inspiration. Any advice? Is this something that is doable?

**Comment by oguiza** [comment]
Hi @Attol8,

Thanks for your proposal to work on the Temporal Fusion Transformer model. I'd love to have it added to `tsai` 😃. I can help you in the process. 

To build a model it's more important to have some experience with Pytorch than fastai. A model can be implemented in pure Pytorch, and they'd work fine in tsai, although you may create a more compact implementation using fastai & tsai layers for example. 

I don't know how familiar you are with transformers. Anyway, to code something is a great opportunity to learn if you are willing to make the effort. And TFT won't be easy, but is definitely doable.  

To create a new implementation in tsai, what I usually do is the following: 

1. Read the paper to understand the input, the model architecture and the output. 
2. Try to find the author's implementation. If it's in Pytorch great. Unfortuanately in this case, it's available in Tensorflow. It may still be useful if you can code in tf.
3. If the authors' implementation is not available in Pytorch, look for an alternative good quality implementation in paperswithcode. It seems that the best implementation might be this https://github.com/jdb78/pytorch-forecasting/tree/master/pytorch_forecasting/models/temporal_fusion_transformer. 
4. A good Pytorch implementation makes your life easier. The next step is to get hold of a dataset. There are 4 datasets they reference in the paper. Try to get hold of any of them. There may be some functionality to download it in the pytorch-forecasting website, I don't know. 
5. If you have the architecture and the data, you can then manually create a batch. What you want to do is to understand what happens to the input in every layer in the model. You may want to use hooks or print out the shape of each layer's output. I always learn a lot when I do this. 
6. There is a TFT tutorial in pytorch-forecasting that might be useful (https://pytorch-forecasting.readthedocs.io/en/latest/tutorials/stallion.html). I haven't read it. 
7. A critical thing to understand to add any model to tsai is the data input the model needs. This may require a modification to the dataloaders, or the use of a callback, creation of masks, etc. We need to understand what data needs to go in each mini-batch. There are many arguments in TFT. We need to understand each one of them.
8. I can help with any changes to the dataloaders or callbacks (if needed), so you can focus on the architecture itself.
9. We'd also need to agree on the models API. I've consistently used a naming convention, and placed the args/ kwargs in a way that makes it easier for the user to swap one architecture by another. I'm not sure how much we'd be able to preserve this common API with TFT.
10. Then, we can compare the individual modules. Some of them might already be available in `tsai`. Some may need to be adapted. And some might just be reused (with proper citation). There shouldn't be any issues because the license is similar to `tsai's`. 

The ideal output of this work is this: 

1. Add one or multiple data sources to tsai to be able to easily download the data.
2. Updated tsai core functionality (if required): dataloders, callback, etc.
3. TFT architecture
4. A tutorial notebook demonstrating how to use it
5. Documentation on the model
6. Set of tests that will be used

I'm not sure if you are aware, but I've used nbdev to build tsai. In my opinion, it works fine. It helps you code, document and test everything.

I recently updating the CONTRIBUTING.md file with a step-by-step guide on how to set up an environment to develop tsai. You can ask me any questions you have about this. 

So that's pretty much how I approach a new architecture. Sometimes they are very simple, and just take a couple of hours, sometimes 2 or more weeks. This one is not particularly easy, although a Pytorch implementation is already available.

Please, let me know if you are still interested. If so, I'll be happy to help you throughout the process. I'm sure we'll both learn a lot!

Thanks,
Ignacio

**Comment by Attol8** [comment]
Hi @oguiza,

Looking forward to implementing this architecture, I believe we can have much fun and learn a lot at the same time. I have read the contribution guide, and your advices, thanks for providing all these details. 

I have a general question: how does a `tsai` implementation differ from a Pytorch one (e.g. The `Pytorch forecasting` implementation that you linked https://github.com/jdb78/pytorch-forecasting/tree/master/pytorch_forecasting/models/temporal_fusion_transformer)? What advantages does such implementation propose?

**Comment by oguiza** [comment]
THi @Attol8,

Well the only thing that needs to be different is the model's API. I'll try to explain it. 

One of the issues I've always found when I try to use state-of-the-art architectures with my own datasets is that each researcher creates a different API. And it's difficult to know what does where. So you need to end up using their own boilerplate and adapt your dataset. That's pretty time consuming and innefficient. 

What I've done with `tsai` is to harmonize the API of multiple SOTA architecures. In `tsai` it's very easy to try different architectures on your dataset, just by replacing the architecture name. 

```
from tsai.all import *
X, y, splits = get_classification_data('ECG200', split_data=False)
batch_tfms = TSStandardize()
clf = TSClassifier(X, y, splits=splits, arch=InceptionTime, batch_tfms=batch_tfms, metrics=accuracy, cbs=ShowGraph(), verbose=True)
clf.fit_one_cycle(100, 3e-4)
```
```
from tsai.all import *
X, y, splits = get_classification_data('ECG200', split_data=False)
batch_tfms = TSStandardize()
clf = TSClassifier(X, y, splits=splits, arch=TST, batch_tfms=batch_tfms, metrics=accuracy, cbs=ShowGraph(), verbose=True)
clf.fit_one_cycle(100, 3e-4)
```

To make this work the essential points are: 
1. What's the expected input to the model
2. Adapt the API so that models can be build using the information in the dataloader. 

So if we re-use a Pytorch model, from the model's perspective, we only need to adapt the API. From a dataloader's perspective, we may need more changes to generate the expected model input. 

By using this approach, we may be able to benchmark different forecasting models. If we have N-BEATS, TFT and Informer, we could compate them (which is something I haven't seen- although I may have missed it). 

I'm not sure if this answers your question.

**Comment by oguiza** [comment]
Hi @Attol8,

I'd love to have an update on the status of this initiative. Are you still interested in implementing TFT in `tsai`?

**Comment by oguiza** [comment]
I'll close this issue due to a lack of response. 
If @Attol8 or anybody else wishes to work on this, please, feel free to reopen.

**Comment by oxcabe** [comment]
Hi, I'm interested in understanding this one, and will take a shot in implementing it. Ill ping you @oguiza in case I get stuck on something. See you in the PR!

**Comment by oguiza** [comment]
> 

Hi @oxcabe,
It sounds great to me. Please, let me know if you have any questions regarding this implementation or the tsai library.
